Untested Files (Buffer)/HLD.cpp

```diff
@@ -44,11 +44,11 @@ class SegmentTree {
 	}
 	inline void propagate(int v, int tl, int tr) {
 		if (lazy[v]) {
-			tree[v] += lazy[v];
+			tree[v] = merge(tree[v],lazy[v]);
 		}
 		if (tl != tr) {
-			lazy[left(v)] += lazy[v];
-			lazy[right(v)] += lazy[v];
+			lazy[left(v)] = merge(lazy[left(v)],lazy[v]);
+			lazy[right(v)] = merge(lazy[right(v)],lazy[v]);
```

Comment:
It might be a better idea to have a different merge function for lazy updates. Something like `lazy_merge`().
Mainly because often the lazy update has a different merging scheme than the query merging itself. (For example, in a seg tree to compute max, lazy updates are for addition, but queries involve taking the maximum.